### PR TITLE
Store migration support for block devices

### PR DESCRIPTION
**Pull request by ragadeeshu**
Make store migration work for custom IO integrations, by accessing store files through the page cache. Including for file operations such as deletes, copying and moving files.

This work includes a new `StoreMigrationIT` test, which searches for all available store formats, and tests migration for all supported from-to format tuples. This test is used in the block device integration code to verify that migration works there as well.

The page cache based file operations only take effect for formats from 2.3 and onward.

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java]: I don't see how there can still be a mapping on any file here. All of the paged files should be closed once the import finishes.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java:261]: Here we delete files through `pageCache` first and `fileSystem` second.
In `StoreMigrator` it is the other way around. 
I think it make sense to use the same pattern between the two. Especially since most difficulties around adapting Neo to behave with blockdevice is around file operations.

Maybe consider using `FileHandle.HANDLE_DELETE`?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java]: We get this `IOException` from `SingleFilePageSwapperFactory.streamFilesRecursiveInner(...)` if `migrationDirectory` does not exist. I would prefer if we specifically caught the `NoSuchFileException` that it throws instead to not hide other possible `IOExceptions`, for example if some file is already mapped we get `FileIsMappedException`.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:499]: Same comment as for `StoreUpgrader`, even though we explicitly `close` mapped files, there could be other `IOException` that we do not want to hide.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:595]: Are all of these calls here to make us fail fast if some store is present? In that case 👍

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:736]: Same `IOException` comment again.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:715]: I like this pattern of first trying to finish the job with `FileSystemAbstraction` and then "clean up" what's left with `PageCache`.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:719]: This block has a lot in common with same block in `migrateWithBatchImporter` when deleting files from `migrationDir`. Maybe the behaviour could be extracted into a helper method by extracting the action (delete, rename) and a `Supplier<Stream<StoreFile>>` to test the predicate for.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java]: Can we please move this assertion to before calling this method and add a comment about it, like we do in `rebuildCounts` so it becomes:
```
if ( versionToMigrateFrom.equals( StandardV2_1.STORE_VERSION ) )
{
    // These versions are not supported for block devices.
    CustomIOConfigValidator.assertCustomIOConfigNotUsed( config, CUSTOM_IO_EXCEPTION_MESSAGE );
    removeDuplicateEntityProperties( storeDir, migrationDir, pageCache, schemaIndexProvider, oldFormat );
}
```

Also, why do we not check for `StandardV2_2.STORE_VERSION` like we do in `rebuildCounts`?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java:120]: Instead of removing this from here and check it later in `StoreMigrator`, does it make sense to instead fail as fast as possible here as soon and `versionToMigrateFrom` is known?
I mean inside the `if ( MigrationStatus.X.isNeededFor( migration status ) )` blocks.

File: community/neo4j/src/test/java/upgrade/PlatformConstraintStoreUpgradeTest.java
Comment (burqen) [inline community/neo4j/src/test/java/upgrade/PlatformConstraintStoreUpgradeTest.java:93]: Should we not also check for `StandardV2_1.STORE_VERSION` here?

**Review by burqen — CHANGES_REQUESTED** [review]
First review round finished. Have some concerns about Exception handling and some small comments to improve readability.

**Comment by burqen** [comment]
@ragadeeshu Comments ready for you 👍

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java
Comment (ragadeeshu) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:595]: These calls are there in order to explicitly create the stores, I think. This method has just been migrated from `BatchingNeoStores` since only this is the only class that calls it.

Comment (burqen) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:595]: Alright, that was my first thought as well but i couldn't see any creation being done in those code paths.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java
Comment (ragadeeshu) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java:120]: One argument against this is that the `StoreUpgrader` would need to become aware of store versions, instead of that being isolated to `StoreMigrator`

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java
Comment (ragadeeshu) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:719]: This, along with consolidating all/most of our file helper classes, should probably be done separately.

File: community/neo4j/src/test/java/upgrade/PlatformConstraintStoreUpgradeTest.java
Comment (ragadeeshu) [reply community/neo4j/src/test/java/upgrade/PlatformConstraintStoreUpgradeTest.java:93]: Fixed, also added 2.0

**Comment by ragadeeshu** [comment]
@burqen Please have a look again :)

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java
Comment (burqen) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:719]: I was thinking of extracting it only for this class, but a larger refactoring might be a good idea. If you plan to do this, then nevermind, otherwise feel free to address it for this class in this PR.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java
Comment (burqen) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java:120]: It's actually `StoreUpgrader` that reads store version and passes it to `StoreMigrator`.

Comment (ragadeeshu) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java:120]: @burqen I meant checking against specific versions, importing `StandardV` classes, etc. `StoreUpgrader` is currently free of such logic, while `StoreMigrator` is full of it already. `StoreMigrater` is also the only `Migrator` implementation that is incompatible with custom IO configurations.

Comment (burqen) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java:120]: Ah, then would agree that it makes more sense keeping it as it is.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java]: I missed this during my first review apparently, sorry.
We should work towards using `FileSystemAbstraction` whenever possible, rather than mixing between `FSA`, `FileUtils` and `File#...`. Feel free to use `fileSystem.fileExists( migrationDirectory )` here.

**Review by burqen — CHANGES_REQUESTED** [review]
Second review finished. Changes look good. Approved once I get an answer on the remaining questions.

**Comment by burqen** [comment]
@ragadeeshu Second review finished. There are some test failures that seem related to these changes as well.

**Review by burqen — APPROVED** [review]

**Pull request merged**
